File: ts/src/knowledge/campaign-mode-report-store.ts
import { existsSync, mkdirSync, readdirSync, readFileSync, statSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import {
  campaignModeReportToMarkdown,
  parseCampaignModeReport,
  type CampaignModeReport,
} from "../analytics/campaign-mode-report.js";

export function campaignModeReportPath(
  knowledgeRoot: string,
  scenarioName: string,
  runId: string,
): string {
  return join(knowledgeRoot, scenarioName, "campaign_mode_reports", `${runId}.json`);
}

export function writeCampaignModeReport(
  knowledgeRoot: string,
  scenarioName: string,
  runId: string,
  report: CampaignModeReport,
): string {
  const path = campaignModeReportPath(knowledgeRoot, scenarioName, runId);
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, JSON.stringify(report, null, 2) + "\n", "utf-8");
  return path;
}

export function readCampaignModeReport(
  knowledgeRoot: string,
  scenarioName: string,
  runId: string,
): CampaignModeReport | null {
  const path = campaignModeReportPath(knowledgeRoot, scenarioName, runId);
  return existsSync(path)
    ? parseCampaignModeReport(JSON.parse(readFileSync(path, "utf-8")) as unknown)
    : null;
}

export function readLatestCampaignModeReportsMarkdown(
  knowledgeRoot: string,
  scenarioName: string,
  opts: { maxReports?: number } = {},
): string {
  const dir = join(knowledgeRoot, scenarioName, "campaign_mode_reports");
  if (!existsSync(dir)) return "";
  return readdirSync(dir)
    .filter((name: string) => name.endsWith(".json"))
    .map((name: string) => join(dir, name))
    .sort((left: string, right: string) => statSync(right).mtimeMs - statSync(left).mtimeMs)
    .slice(0, opts.maxReports ?? 2)
    .map((path: string) =>
      campaignModeReportToMarkdown(
        parseCampaignModeReport(JSON.parse(readFileSync(path, "utf-8")) as unknown),
      ),
    )
    .join("\n\n");
}

File: autocontext/src/autocontext/storage/campaign_mode_report_store.py
"""File helpers for campaign-mode reports."""

from __future__ import annotations

from pathlib import Path, PurePosixPath, PureWindowsPath
from typing import Any, Protocol

from autocontext.analytics.campaign_mode_report import CampaignModeReport
from autocontext.storage.scenario_paths import normalize_scenario_name_segment
from autocontext.util.json_io import read_json_guarded, write_json


class DictSerializable(Protocol):
    def to_dict(self) -> dict[str, Any]: ...


def campaign_mode_report_path(knowledge_root: Path, scenario_name: str, run_id: str) -> Path:
    return (
        knowledge_root
        / normalize_scenario_name_segment(scenario_name)
        / "campaign_mode_reports"
        / f"{_normalize_path_segment(run_id, 'run_id')}.json"
    )


def _normalize_path_segment(value: str, label: str) -> str:
    normalized = value.strip()
    if not normalized:
        raise ValueError(f"{label} is required")
    if "/" in normalized or "\\" in normalized:
        raise ValueError(f"{label} must be a single path segment: {value!r}")
    for path_cls in (PurePosixPath, PureWindowsPath):
        candidate = path_cls(normalized)
        if candidate.is_absolute() or len(candidate.parts) != 1 or candidate.parts[0] in {".", ".."}:
            raise ValueError(f"{label} must be a single path segment: {value!r}")
    return normalized


def write_campaign_mode_report(knowledge_root: Path, scenario_name: str, run_id: str, report: DictSerializable) -> Path:
    path = campaign_mode_report_path(knowledge_root, scenario_name, run_id)
    write_json(path, report.to_dict())
    return path


def read_campaign_mode_report_path(path: Path) -> CampaignModeReport | None:
    """Parse one report file; corrupt files degrade to None."""
    data = read_json_guarded(path)
    if not isinstance(data, dict):
        return None
    try:
        return CampaignModeReport.from_dict(data)
    except (TypeError, ValueError):
        return None


def read_campaign_mode_report(knowledge_root: Path, scenario_name: str, run_id: str) -> CampaignModeReport | None:
    path = campaign_mode_report_path(knowledge_root, scenario_name, run_id)
    if not path.exists():
        return None
    data = read_json_guarded(path)
    if not isinstance(data, dict):
        return None
    try:
        return CampaignModeReport.from_dict(data)
    except (TypeError, ValueError):
        return None


def read_latest_campaign_mode_reports_markdown(
    knowledge_root: Path,
    scenario_name: str,
    *,
    max_reports: int = 2,
) -> str:
    root = knowledge_root / normalize_scenario_name_segment(scenario_name) / "campaign_mode_reports"
    if not root.exists():
        return ""
    paths = sorted(root.glob("*.json"), key=lambda path: path.stat().st_mtime, reverse=True)[:max_reports]
    parts: list[str] = []
    for path in paths:
        report = read_campaign_mode_report_path(path)
        if report is not None:
            parts.append(report.to_markdown())
    return "\n\n".join(parts)


__all__ = [
    "campaign_mode_report_path",
    "read_campaign_mode_report",
    "read_latest_campaign_mode_reports_markdown",
    "write_campaign_mode_report",
]
